### ISSUE TITLE
System.Security.Cryptography.Native.OpenSsl not found on Ubuntu 14.04

### ISSUE THREAD

**Issue by moozzyk** [dotnet/corefx#12517]
A SignalR sample project fails on Linux because it cannot find System.Security.Cryptography.Native.OpenSsl. I turned on LD_DEBUG and can see that it is probing in different location but is not able to find it. SignalR uses "System.Security.Cryptography.Algorithms": "4.2.0-*" to [get a random number](https://github.com/aspnet/SignalR-Server/blob/9680e9c2c185a19604390003a6f10489d0538a56/src/Microsoft.AspNetCore.SignalR.Sources/SipHashBasedStringEqualityComparer.cs#L46)

```
     27185:     file=/usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/System.Security.Cryptography.Native.OpenSsl [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=/home/xxxxxx/.nuget/packages/System.Security.Cryptography.Algorithms/4.3.0-preview1-24530-03/runtimes/unix/lib/netstandard1.6/System.Security.Cryptography.Native.OpenSsl [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=System.Security.Cryptography.Native.OpenSsl [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:     find library=System.Security.Cryptography.Native.OpenSsl [0]; searching
     27185:      search cache=/etc/ld.so.cache
     27185:      search path=/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib:/usr/lib              (system search path)
     27185:       trying file=/lib/x86_64-linux-gnu/System.Security.Cryptography.Native.OpenSsl
     27185:       trying file=/usr/lib/x86_64-linux-gnu/System.Security.Cryptography.Native.OpenSsl
     27185:       trying file=/lib/System.Security.Cryptography.Native.OpenSsl
     27185:       trying file=/usr/lib/System.Security.Cryptography.Native.OpenSsl
     27185:
     27185:
     27185:     file=/home/xxxxxx/.nuget/packages/System.Security.Cryptography.Algorithms/4.3.0-preview1-24530-03/runtimes/unix/lib/netstandard1.6/libSystem.Security.Cryptography.Native.OpenSsl.so [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=/usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libSystem.Security.Cryptography.Native.OpenSsl.so [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=libSystem.Security.Cryptography.Native.OpenSsl.so [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:     find library=libSystem.Security.Cryptography.Native.OpenSsl.so [0]; searching
     27185:      search cache=/etc/ld.so.cache
     27185:      search path=/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib:/usr/lib              (system search path)
     27185:       trying file=/lib/x86_64-linux-gnu/libSystem.Security.Cryptography.Native.OpenSsl.so
     27185:       trying file=/usr/lib/x86_64-linux-gnu/libSystem.Security.Cryptography.Native.OpenSsl.so
     27185:       trying file=/lib/libSystem.Security.Cryptography.Native.OpenSsl.so
     27185:       trying file=/usr/lib/libSystem.Security.Cryptography.Native.OpenSsl.so
     27185:
     27185:
     27185:     file=/home/xxxxxx/.nuget/packages/System.Security.Cryptography.Algorithms/4.3.0-preview1-24530-03/runtimes/unix/lib/netstandard1.6/System.Security.Cryptography.Native.OpenSsl.so [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=/usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/System.Security.Cryptography.Native.OpenSsl.so [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=System.Security.Cryptography.Native.OpenSsl.so [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:     find library=System.Security.Cryptography.Native.OpenSsl.so [0]; searching
     27185:      search cache=/etc/ld.so.cache
     27185:      search path=/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib:/usr/lib              (system search path)
     27185:       trying file=/lib/x86_64-linux-gnu/System.Security.Cryptography.Native.OpenSsl.so
     27185:       trying file=/usr/lib/x86_64-linux-gnu/System.Security.Cryptography.Native.OpenSsl.so
     27185:       trying file=/lib/System.Security.Cryptography.Native.OpenSsl.so
     27185:       trying file=/usr/lib/System.Security.Cryptography.Native.OpenSsl.so
     27185:
     27185:
     27185:     file=/home/xxxxxx/.nuget/packages/System.Security.Cryptography.Algorithms/4.3.0-preview1-24530-03/runtimes/unix/lib/netstandard1.6/libSystem.Security.Cryptography.Native.OpenSsl [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=/usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libSystem.Security.Cryptography.Native.OpenSsl [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:
     27185:     file=libSystem.Security.Cryptography.Native.OpenSsl [0];  dynamically loaded by /usr/share/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so [0]
     27185:     find library=libSystem.Security.Cryptography.Native.OpenSsl [0]; searching
     27185:      search cache=/etc/ld.so.cache
     27185:      search path=/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib:/usr/lib              (system search path)
     27185:       trying file=/lib/x86_64-linux-gnu/libSystem.Security.Cryptography.Native.OpenSsl
     27185:       trying file=/usr/lib/x86_64-linux-gnu/libSystem.Security.Cryptography.Native.OpenSsl
     27185:       trying file=/lib/libSystem.Security.Cryptography.Native.OpenSsl
     27185:       trying file=/usr/lib/libSystem.Security.Cryptography.Native.OpenSsl
```

```
System.TypeInitializationException: The type initializer for 'Crypto' threw an exception. ---> System.DllNotFoundException: Unable to load DLL 'System.Security.Cryptography.Native.OpenSsl': The specified module could not be found.
 (Exception from HRESULT: 0x8007007E)
   at Interop.Crypto.GetMaxMdSize()
   at Interop.Crypto..cctor()
   --- End of inner exception stack trace ---
   at Interop.Crypto.GetRandomBytes(Byte[] buf, Int32 num)
   at System.Security.Cryptography.RNGCryptoServiceProvider.GetBytes(Byte[] data)
   at Microsoft.AspNetCore.SignalR.Infrastructure.SipHashBasedStringEqualityComparer.GenerateRandomKeySegment()
   at Microsoft.AspNetCore.SignalR.Infrastructure.SipHashBasedStringEqualityComparer..ctor()
   at Microsoft.AspNetCore.SignalR.Messaging.TopicLookup..ctor()
   at Microsoft.AspNetCore.SignalR.Messaging.MessageBus..ctor(IStringMinifier stringMinifier, ILoggerFactory loggerFactory, IPerformanceCounterManager performanceCounterManager, IOptions`1 optionsAccessor)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.AspNetCore.SignalR.PersistentConnection.Initialize(IServiceProvider serviceProvider)
   at Microsoft.AspNetCore.SignalR.Hubs.HubDispatcher.Initialize(IServiceProvider serviceProvider)
   at Microsoft.AspNetCore.SignalR.Hosting.PersistentConnectionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.Internal.RequestServicesContainerMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.Frame`1.<RequestProcessingAsync>d__2.MoveNext()
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
```

**Comment by karelz** [comment]
@bartonjs @steveharter is it a known issue?

**Comment by moozzyk** [comment]
See @halter73 comment: https://github.com/aspnet/SignalR-Server/issues/237#issuecomment-252751296

**Comment by halter73** [comment]
Could the System.Security.Cryptography.Native.OpenSsl.so shim go into a fat package?

I think it's a really bad experience to have an app the successfully restores and builds without errors on all platforms and runs just fine on Windows, but blows up at runtime on Linux and Mac.

**Comment by moozzyk** [comment]
Ideally this kinds of conditions should be detected at design time (e.g. during restore) since they are hard to diagnose. Also, the application works just fine on Windows and Mac.

**Comment by steveharter** [comment]
@ericstj @bartonjs question\issue on packaging the native shims. It appears the .OpenSsl package was added in August via https://github.com/dotnet/corefx/commit/45172b7136f32873ff60880ebe654a6226895542#diff-2566e1ae9c5bac4fe163bdfc4ce50b53 as part of the work to support non-OpenSsl scenarios on Apple. 

Thus references to older\mismatched "Microsoft.NETCore.App" is not finding the newer .OpenSsl package. 

Can we detect this condition more gracefully?

**Comment by ericstj** [comment]
> Can we detect this condition more gracefully? 

We shouldn't change the implementation used for netcoreapp1.0 to depend on a shim update.  Netcoreapp1.0 restores without a RID, so it must have assets that can be satisfied by RID-specific shared framework.  @bartonjs I thought we discussed this scenario?  Perhaps we missed a case with latest packages running on netcoreapp1.0.

**Comment by ericstj** [comment]
@moozzyk can you confirm this is a netcoreapp1.0 project using 1.0 shared framework with the latest packages?

**Comment by moozzyk** [comment]
```
  "frameworks": {
    "net451": {},
    "netcoreapp1.0": {
      "dependencies": {
        "Microsoft.NETCore.App": {
          "version": "1.0.1-*",
          "type": "platform"
        },
        "System.Runtime.Serialization.Primitives": "4.3.0-*"
      }
    }
  },
```
but the project uses: 
Microsoft.AspNetCore.SignalR.Messaging which depends on "NETStandard.Library": "1.6.1-*" (changed recently)

**Comment by moozzyk** [comment]
Here is the project: https://github.com/aspnet/SignalR-Server/tree/dev/samples/SignalRSample.Web - only Microsoft.NETCore.App version was changed to 1.1.0.

**Comment by ericstj** [comment]
Yeah, so you're using 1.1 packages on the 1.0 shared framework.  That does explain this.

I see 3 ways to fix it:

1. Cross-compile (or redist) crypto packages for netcoreapp1.0/1.1 and keep the old name in the 1.0 IL.
2. Undo the rename, or at least fallback to the old name.
3. Make native shim package have direct dependencies on split packages instead of runtime dependencies, similar to what we already do for SNI.dll.

**Comment by bartonjs** [comment]
We can't just (un)do the rename, because the whole reason the name changed was to support SxS with using more of Apple's crypto libraries (reducing the OpenSSL dependencies on macOS for 1.1, with the goal of eliminating them (aside from people explicitly requesting (e.g.) RSAOpenSsl) in 1.2).  While the rename of the S.S.C.Native library to S.S.C.Native.OpenSsl isn't strictly required, if the title of this bug was "System.Security.Cryptography.Native.Apple was not found on osx.10.11" we'd have come to different initial realizations.

If we went redist, we'd no longer be automagically backfixing a user reported bugfix (#8712), and we'd also not be backporting the perf gains of #9510.

At least one function was added to the native shim between 1.0.0 and 1.1.0: #9993 added a new function to the shim to reduce the memory overhead in preventing early disposing.

So, if I understand the state of the world correctly, redist is fine, and direct dependencies is fine... whatever seems the best approach for 1.1 at this point is fine with me.  And if I didn't understand the state of the world correctly hopefully I gave enough data to help someone who does correctly steer the valid options set :smile:.

**Comment by bartonjs** [comment]
Testing the fix

**Comment by bartonjs** [comment]
This was fixed by #13060.

**Comment by karelz** [comment]
@jiangzeng I don't think this is something we would hit in our test runs. If you are hitting similar exceptions in test runs, please file a new bug, thanks!
